src/hivpy/prep.py

```diff
@@ -356,3 +411,119 @@ def prep_eligibility(self, pop: Population):
+            prep_types = pop.transform_group([col.PREP_ORAL_RANK, col.PREP_CAB_RANK,
+                                              col.PREP_LEN_RANK, col.PREP_VR_RANK,
+                                              col.PREP_ORAL_WILLING, col.PREP_CAB_WILLING,
+                                              col.PREP_LEN_WILLING, col.PREP_VR_WILLING],
+                                             self.calc_willing_start_prep, sub_pop=starting_prep_pop)
+
+            pop.set_present_variable(col.PREP_TYPE, prep_types, starting_prep_pop)
+            pop.set_present_variable(col.EVER_PREP, True, starting_prep_pop)
+            pop.set_present_variable(col.LAST_PREP_START_DATE, pop.date, starting_prep_pop)
+
+            def set_prep_start_date(pop: Population, starting_prep_pop, prep_type, start_date_col):
+                """
+                Set a specific start date column for the population starting a corresponding PrEP type.
+                """
+                pop.set_present_variable(start_date_col, pop.date,
+                                         pop.get_sub_pop_intersection(
+                                             starting_prep_pop,
+                                             pop.get_sub_pop(COND(col.PREP_TYPE, op.eq, prep_type))))
+
+            set_prep_start_date(pop, starting_prep_pop, PrEPType.Oral, col.FIRST_ORAL_START_DATE)
+            set_prep_start_date(pop, starting_prep_pop, PrEPType.Cabotegravir, col.FIRST_CAB_START_DATE)
+            set_prep_start_date(pop, starting_prep_pop, PrEPType.Lenacapavir, col.FIRST_LEN_START_DATE)
+            set_prep_start_date(pop, starting_prep_pop, PrEPType.VaginalRing, col.FIRST_VR_START_DATE)
+
+    def calc_willing_start_prep(self, oral_pref, cab_pref, len_pref, vr_pref,
+                                oral_willing, cab_willing, len_willing, vr_willing, size):
+        """
+        Returns PrEP types for people starting PrEP for the first time without explicitly
+        testing to start PrEP. Individual preferences and availability are taken into account.
+        """
+        # group pref ranks and willingness
+        prefs = [oral_pref, cab_pref, len_pref, vr_pref]
+        willing = [oral_willing, cab_willing, len_willing, vr_willing]
+        # zip prep type and willingness together and sort by pref rank
+        sorted_zipped = sorted(enumerate(willing), key=lambda x: prefs[x[0]])
+        sorted_dict = dict(sorted_zipped)
+
+        starting_prep = None
+        # find prep type someone is willing to take with the highest pref that is currently available
+        for prep_type in sorted_dict:
+            willing = sorted_dict[prep_type]
+            if self.date >= self.date_prep_intro[prep_type] and willing:
+                starting_prep = prep_type
+                break
+
+        # outcomes
+        r = rng.uniform(size=size)
+        # FIXME: may need different probabilities for different prep types
+        starting = r < self.prob_base_prep_start
```

Comment:
It's been confirmed that we want to retain the ability to apply different probabilities based on PrEP type, so this should be investigated in this PR as well.